presto-main/src/main/java/io/prestosql/operator/HttpPageBufferClient.java

```diff
@@ -605,6 +639,33 @@ public PagesResponse handle(Request request, Response response)
             }
         }
 
+        private void verifyChecksum(URI uri, long readChecksum, List<SerializedPage> pages)
+        {
+            // TODO identify our address, so that query failure reports both ends of communication
+            switch (dataIntegrityVerification) {
+                case NONE:
+                    if (readChecksum != NO_CHECKSUM) {
+                        throw new IllegalStateException(format("Expected checksum to be NO_CHECKSUM (0x%08x) but is 0x%08x, when reading from %s", NO_CHECKSUM, readChecksum, uri));
+                    }
+                    return;
+
+                case ABORT:
+                case RETRY:
+                    long calculatedChecksum = calculateChecksum(pages);
+                    if (readChecksum != calculatedChecksum) {
+                        String message = format("Data corruption, read checksum: 0x%08x, calculated checksum: 0x%08x, when reading from %s", readChecksum, calculatedChecksum, uri);
+                        if (dataIntegrityVerification == DataIntegrityVerification.RETRY) {
+                            throw new IllegalStateException(message);
+                        }
+                        throw new PrestoException(GENERIC_INTERNAL_ERROR, message);
```

Comment:
This is brittle. We need a better mechanism to indicate we want to retry than the type of exception. This logic to check for retry policy should really be in the code that performs the actual retries, otherwise there's a high chance it will get out of sync in the future.

We should throw a non-retryable exception in the case of configuration or a "checksum failed" exception otherwise. The code that does retries in HttpPageBufferClient should check the retry policy and act accordingly in case of "checksum failed".

Reply:
Updated the retry code.